src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp

```diff
@@ -0,0 +1,79 @@
+
+#include "software/geom/algorithms/closest_point.h"
+
+bool PassDefenderFSM::passStarted(const Update& event)
+{
+    auto ball_position = event.common.world.ball().position();
+    Vector ball_receiver_point_vector(
+        event.control_params.position_to_block_from.x() - ball_position.x(),
+        event.control_params.position_to_block_from.y() - ball_position.y());
+
+    bool pass_started = event.common.world.ball().hasBallBeenKicked(
+        ball_receiver_point_vector.orientation(), MIN_PASS_SPEED,
+        MAX_PASS_ANGLE_DIFFERENCE);
+
+    if (pass_started)
+    {
+        // We want to keep track of the initial trajectory of the pass
+        // so that we can later tell whether the ball strays from
+        // this trajectory
+        pass_orientation = event.common.world.ball().velocity().orientation();
+    }
+
+    return pass_started;
+}
+
+bool PassDefenderFSM::ballDeflected(const Update& event)
+{
+    auto orientation_difference =
+        event.common.world.ball().velocity().orientation().minDiff(pass_orientation);
+
+    // If the ball strays from the initial trajectory of the pass,
+    // it was likely deflected off course by another robot or chipped
+    // away by the pass defender
+    return orientation_difference.abs() > MIN_DEFLECTION_ANGLE;
+}
+
+void PassDefenderFSM::blockPass(const Update& event)
+{
+    auto position_to_block_from = event.control_params.position_to_block_from;
+    auto ball_position          = event.common.world.ball().position();
+    auto face_ball_orientation =
+        (ball_position - event.common.robot.position()).orientation();
+
+    // Face the ball and move to position_to_block_from, which should be a location
+    // on the field that blocks a passing lane between two enemy robots
+    event.common.set_primitive(createMovePrimitive(
+        CREATE_MOTION_CONTROL(position_to_block_from), face_ball_orientation, 0,
+        TbotsProto::DribblerMode::OFF, TbotsProto::BallCollisionType::ALLOW,
+        AutoChipOrKick{AutoChipOrKickMode::OFF, 0},
+        TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0,
+        event.common.robot.robotConstants()));
+}
+
+void PassDefenderFSM::interceptBall(const Update& event)
+{
+    auto ball           = event.common.world.ball();
+    auto robot_position = event.common.robot.position();
+
+    if ((ball.position() - robot_position).length() >
+        BALL_TO_FRONT_OF_ROBOT_DISTANCE_WHEN_DRIBBLING)
+    {
+        // Find the closest point on the line of the ball's current trajectory
+        // that the defender can move to and intercept the pass
+        auto intercept_position = closestPoint(
+            robot_position, Line(ball.position(), ball.position() + ball.velocity()));
+
+        auto face_ball_orientation = (ball.position() - robot_position).orientation();
+
+        // Move to intercept the pass by positioning defender in front of the
+        // ball's current trajectory
```

Comment:
Thanks for the insight, some notes:

- It's think fine if we encounter the scenarios you mentioned since our defender likely won't be able to intercept the ball in time anyway. Given the speed of the game, it's better to "fail" first and quickly realize the intercept isn't going to work, then return back to the lane blocking state. Otherwise we get behaviours like ball chasing and this puts our defenders out of position, creating a lot of open space on the field and allowing the enemy team to set up again. This is a known weakness that good teams take advantage of (e.g. Tigers, CMDragons in the past)

- The pass defender should already be blocking the shooting lane, so this intercept function is really only supposed to make small adjustments to the positioning of the defender to match the trajectory of the moving ball. If it's wildly out of position intercepting the pass was probably out of the question, so now it's the goalie and crease defenders' job to stop the ball from entering the goal.

- We should implement some additional guards to encourage the defender to exit the intercept state if it's clear that the intercept will fail (ex. guard that checks if the ball is moving further away from the defender)

- There may be issues with having one common interception algorithm shared between all the tactics given the different play styles in defense and offense. I think this avenue has been explored before (#1968) but it's worth reinvestigating. This is probably outside of the scope of this pr so I can open a new issue

Reply:
I see. You bring up great points. I guess we will have to make sure that these various scenarios are well tested so we can determine whether extra guards are needed or not.